docs/design/2025-03-17-sys-session-manage-enhancement.md

```diff
@@ -0,0 +1,173 @@
+# Enhancement of Internal System Session Management
+
+- Author: [Wang Chao](https://github.com/lcwangchao)
+- Discussion PR: <https://github.com/pingcap/tidb/pull/59875>
+- Tracking Issue: <https://github.com/pingcap/tidb/issues/60115>
+
+## Motivation and Background
+
+The internal system sessions are widely used in TiDB across multiple modules. To minimize the overhead of creating and destroying sessions, we use a session pool for management. However, we have recently encountered several issues with the current session pool implementation—most of which stem from its improper usage.
+
+For example, in issue [#56934](https://github.com/pingcap/tidb/issues/56934), the TTL module retrieves a session from the pool but fails to return it after use, leading to a memory leak. Another issue arose after PR [#32726](https://github.com/pingcap/tidb/pull/32726), where each session retrieved from the pool is stored as a reference in a map. This reference is only removed when the session is returned to the pool. As a result, if a session is closed using the `Close` method, it may still cause a memory leak because it remains referenced in the map.
+
+Some PR such as [#59546](https://github.com/pingcap/tidb/pull/59546) attempt to address these issues by introducing a `Destroy` method for the session pool. However, this approach does not fully resolve the problems and still makes it difficult to ensure that developers call the correct method.
+
+This document proposes a more effective solution for managing internal system sessions to achieve the following goals:
+
+- Prevent concurrent usage across modules: Once a session is returned to the pool, it should no longer be used to avoid potential conflicts from concurrent access.
+- Simplify session lifecycle management: Ensure that sessions are always either returned to the pool or properly closed after use, eliminating the risk of memory leaks.
+- Enforce session state reset: When a session is returned to the pool, its internal state should be forcibly reset to prevent unexpected behaviors in subsequent usage.
+
+## Design
+
+### Pool
+
+We need to introduce a new enhanced session pool to manage internal system sessions. The new session pool will provide the following methods:
+
+```go
+type Pool struct {
+    ...
+}
+
+func (p *Pool) get() (*Session, error) {
+    ...
+}
+
+func (p *Pool) put(s *Session) {
+    ...
+}
+
+func (p *Pool) WithSession(func(*Session) error) error {
+    ...
+}
+```
+
+The most important method of `Pool` is `WithSession` which is introduced to replace the `Get` and `Put` methods of the current session pool. The `WithSession` method accepts a function as an argument, which will be executed with a session retrieved from the pool. After the function completes, the session will be automatically returned to the pool. This design ensures that the session is always returned to the pool after use, preventing memory leaks.
```

Comment:
> This design ensures that the session is always returned to the pool after use

This can't be strictly ensured, because the function may store the provided session to elsewhere. We still need to use comment of `WithSession` to tell developer not to do it.

Reply:
`WithSession` is designed to avoid forgetting to call `Close` of a session which will never be used again like issue: #56934 . Refering a `Session` in  `WithSession ` elsewhere will indeed cause some memory leak but I think it is not the main problem because:

1. As the `session` returned to the pool, the next `Get` will reuse it without creating a new one.
2. Usually, if `Session` is refered  somewhere by mistake, the amount should not be too much because most of them should be GC by go.